File: xpod/services/XPCompiler.m
//
//  XPCompiler.m
//  xpod
//
//  Created by Maxime Bokobza on 02/06/14.
//  Copyright (c) 2014 dispatchasync. All rights reserved.
//

#import "XPCompiler.h"
#import "XPPod.h"
#import "NSTask+XPPod.h"


NSString * const kLibsDirectory = @"libs";
NSString * const kObjectFilesDirectory = @"object_files";


typedef NS_ENUM(NSUInteger, XPCompilerMode) {
    XPCompilerModeDevice,
    XPCompilerModeSimulator
};


@interface XPCompiler ()

- (void)compileForArch:(NSString *)arch withMinVersionArg:(NSString *)minVersionArg sysroot:(NSString *)sysroot sourceDir:(NSString *)sourceDir name:(NSString *)name;
- (void)linkForArch:(NSString *)arch sysroot:(NSString *)sysroot name:(NSString *)name;

- (NSString *)sdkPathForMode:(XPCompilerMode)mode;
- (NSString *)minVersionArgumentForMode:(XPCompilerMode)mode;
- (NSString *)pathToObjectFilesDirectoryForArch:(NSString *)arch name:(NSString *)name;

@end


@implementation XPCompiler

- (void)compilePod:(XPPod *)pod sourceDirectory:(NSString *)sourceDir podArchitectureOnly:(BOOL)podArchsOnly {
    NSAssert(self.podRootDirectory, @"self.podRootDirectory must not be nil");
    
    NSString *deviceMinVersionArg = [self minVersionArgumentForMode:XPCompilerModeDevice];
    NSString *deviceSDKPath = [self sdkPathForMode:XPCompilerModeDevice];
    
    for (NSString *arch in pod.architectures) {
        [self compileForArch:arch
           withMinVersionArg:deviceMinVersionArg
                     sysroot:deviceSDKPath
                   sourceDir:sourceDir
                        name:pod.name];
        
        [self linkForArch:arch
                  sysroot:deviceSDKPath
                     name:pod.name];
    }
    
    if (!podArchsOnly) {
        NSString *simulatorMinVersionArg = [self minVersionArgumentForMode:XPCompilerModeSimulator];
        NSString *simulatorSDKPath = [self sdkPathForMode:XPCompilerModeSimulator];
        
        for (NSString *arch in @[@"i386", @"x86_64"]) {
            [self compileForArch:arch
               withMinVersionArg:simulatorMinVersionArg
                         sysroot:simulatorSDKPath
                       sourceDir:sourceDir
                            name:pod.name];
            
            [self linkForArch:arch
                      sysroot:simulatorSDKPath
                         name:pod.name];
        }
    }
}

- (void)compileForArch:(NSString *)arch withMinVersionArg:(NSString *)minVersionArg sysroot:(NSString *)sysroot sourceDir:(NSString *)sourceDir name:(NSString *)name {
    NSArray *arguments = @[@"-x", @"objective-c",
                           @"-arch", arch,
                           minVersionArg,
                           @"-fmessage-length=0",
                           @"-fdiagnostics-show-note-include-stack",
                           @"-fmacro-backtrace-limit=0",
                           @"-std=gnu99",
                           @"-fobjc-arc",
                           @"-Wno-trigraphs",
                           @"-fpascal-strings",
                           @"-O0",
                           @"-isysroot", sysroot,
                           @"-c"];
    
    arguments = [arguments arrayByAddingObjectsFromArray:[NSTask filenamesInDirectory:sourceDir ofType:@"m"]];
    
    [NSTask launchTaskFromDirectory:[self pathToObjectFilesDirectoryForArch:arch name:name]
                     withLaunchPath:@"/usr/bin/clang"
                          arguments:arguments];
}

- (void)linkForArch:(NSString *)arch sysroot:(NSString *)sysroot name:(NSString *)name {
    NSArray *arguments = @[@"-static",
                           @"-arch_only", arch,
                           @"-syslibroot", sysroot,
                           @"-framework", @"Foundation"];
    
    NSString *objectFiles = [self pathToObjectFilesDirectoryForArch:arch name:name];
    NSString *libs = [self.podRootDirectory stringByAppendingPathComponent:kLibsDirectory];
    
    arguments = [arguments arrayByAddingObjectsFromArray:[NSTask filenamesInDirectory:objectFiles ofType:@"o"]];
    arguments = [arguments arrayByAddingObjectsFromArray:@[@"-o", [NSString stringWithFormat:@"%@/%@_%@.a", libs, name, arch]]];
    
    [NSTask launchTaskFromDirectory:[self.podRootDirectory stringByAppendingPathComponent:kLibsDirectory]
                     withLaunchPath:@"/usr/bin/libtool"
                          arguments:arguments];
}

- (NSString *)sdkPathForMode:(XPCompilerMode)mode {
    NSString *target = mode == XPCompilerModeDevice ? @"iphoneos" : @"iphonesimulator";
    NSString *output = [NSTask launchTaskWithLaunchPath:@"/usr/bin/xcodebuild"
                                              arguments:@[@"-sdk", target, @"-version"]];
    
    NSRange startRange = [output rangeOfString:@"Path: "];
    output = [output substringFromIndex:startRange.location + startRange.length];
    NSRange endRange = [output rangeOfString:@".sdk"];
    return [output substringToIndex:endRange.location + endRange.length];
}

- (NSString *)minVersionArgumentForMode:(XPCompilerMode)mode {
    // Fetch the right version
    NSString *version = @"7.0";
    NSString *arg = nil;
    switch (mode) {
        case XPCompilerModeDevice:
            arg = @"-miphoneos-version-min=";
            break;
            
        case XPCompilerModeSimulator:
            arg = @"-mios-simulator-version-min=";
            break;
    }
    
    return [arg stringByAppendingString:version];
}

- (NSString *)pathToObjectFilesDirectoryForArch:(NSString *)arch name:(NSString *)name {
    return [[self.podRootDirectory stringByAppendingPathComponent:kObjectFilesDirectory] stringByAppendingFormat:@"/%@_%@", name, arch];
}

@end

File: xpod/services/XPCompiler.swift
//
//  XPCompiler.swift
//  xpod
//
//  Created by Maxime Bokobza on 12/06/14.
//  Copyright (c) 2014 dispatchasync. All rights reserved.
//

import Foundation


let kLibsDirectory: String = "libs";
let kObjectFilesDirectory: String = "object_files";
let kHeadersDirectory: String = "headers";


enum XPCompilerMode: Int {
    case Device, Simulator
}


class XPCompiler {
    var podRootDirectory: String = ""
    
    func compileAndLink(pod: XPPod, sourceDir: String, podArchsOnly: Bool) {
        let deviceMinVersionArg = self.minVersionArgument(XPCompilerMode.Device)
        let deviceSDKPath = self.sdkPath(XPCompilerMode.Device)
        
        for arch: AnyObject in pod.architectures {
            self.compile(arch as String, minVersionArg: deviceMinVersionArg, sysroot: deviceSDKPath, sourceDir: sourceDir, name: pod.name)
            self.link(arch as String, sysroot: deviceSDKPath, name: pod.name)
        }

        if (!podArchsOnly) {
            let simulatorMinVersionArg = self.minVersionArgument(XPCompilerMode.Simulator)
            let simulatorSDKPath = self.sdkPath(XPCompilerMode.Simulator)
            
            for arch in ["i386", "x86_64"] {
                self.compile(arch, minVersionArg: simulatorMinVersionArg, sysroot: simulatorSDKPath, sourceDir: sourceDir, name: pod.name)
                self.link(arch, sysroot: simulatorSDKPath, name: pod.name)
            }
        }
        
        self.copyHeaders(sourceDir)
    }
    
    func compile(arch: String, minVersionArg: String, sysroot: String, sourceDir: String, name: String) {
        var arguments = ["-x", "objective-c",
                        "-arch", arch,
                        minVersionArg,
                        "-fmessage-length=0",
                        "-fdiagnostics-show-note-include-stack",
                        "-fmacro-backtrace-limit=0",
                        "-std=gnu99",
                        "-fobjc-arc",
                        "-Wno-trigraphs",
                        "-fpascal-strings",
                        "-O0",
                        "-isysroot", sysroot,
                        "-c"]
        
        arguments += NSTask.listFiles(sourceDir, fileType: "m")
        
        NSTask.launchTask(self.pathToObjectFilesDirectory(arch, name: name), launchPath: "/usr/bin/clang", arguments: arguments)
    }
    
    func link(arch: String, sysroot: String, name: String) {
        var arguments = ["-static",
                        "-arch_only", arch,
                        "-syslibroot", sysroot,
                        "-framework", "Foundation"]
        
        let objectFiles = self.pathToObjectFilesDirectory(arch, name: name)
        let libs = self.podRootDirectory + "/" + kLibsDirectory
        
        arguments += NSTask.listFiles(objectFiles, fileType: "o")
        arguments += ["-o", "\(libs)/\(name)_\(arch).a"]
        
        NSTask.launchTask(self.podRootDirectory + "/" + kLibsDirectory, launchPath: "/usr/bin/libtool", arguments: arguments)
    }
    
    func copyHeaders(sourceDir: String) {
        let files = NSTask.listFiles(sourceDir, fileType: "h")
        NSTask.copy(files, destination: self.podRootDirectory + "/" + kHeadersDirectory)
    }
    
    func sdkPath(mode: XPCompilerMode) -> String {
        let target = mode == XPCompilerMode.Device ? "iphoneos" : "iphonesimulator"
        var output = NSTask.launchTask("/usr/bin/xcodebuild", arguments: ["-sdk", target, "-version"])

        let regex = NSRegularExpression(pattern: "Path: (.*)sdk", options: NSRegularExpressionOptions.CaseInsensitive, error: nil)
        let result = regex.firstMatchInString(output, options: NSMatchingOptions.ReportProgress, range: NSMakeRange(0, countElements(output)))
        let resultString = (output as NSString).substringWithRange(result.range)
        return resultString.stringByReplacingOccurrencesOfString("Path: ", withString: "")
    }
    
    func minVersionArgument(mode: XPCompilerMode) -> String {
        // Fetch the right version
        let version = "7.0"
        
        switch (mode) {
        case .Device:
            return "-miphoneos-version-min=" + version
            
        case .Simulator:
            return "-mios-simulator-version-min=" + version
        }
    }
    
    func pathToObjectFilesDirectory(arch: String, name: String) -> String {
        return self.podRootDirectory + "/\(kObjectFilesDirectory)/\(name)_\(arch)"
    }
}
